test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -2439,40 +2439,70 @@ class test_ucp_sockaddr_protocols : public test_ucp_sockaddr {
         }
     }
 
+    enum {
+        SEND_PREREG_MEMH = UCS_BIT(0),
+        RECV_PREREG_MEMH = UCS_BIT(1)
+    };
+
+    void save_and_reset_mds(entity &e, std::vector<uct_md_h> &mds)
+    {
+        ucp_tl_md_t *tl_md;
+        ucs_carray_for_each(tl_md, e.ucph()->tl_mds, e.ucph()->num_mds) {
+            mds.push_back(tl_md->md);
+            tl_md->md = NULL;
+        }
+    }
+
+    void restore_mds(entity &e, std::vector<uct_md_h> &mds)
+    {
+        unsigned i = 0;
+        ucp_tl_md_t *tl_md;
+        ucs_carray_for_each(tl_md, e.ucph()->tl_mds, e.ucph()->num_mds) {
+            tl_md->md = mds[i++];
+        }
+    }
+
     void test_am_send_recv(size_t size, size_t hdr_size = 0ul,
                            size_t num_iters = m_num_iters,
-                           bool recv_prereg_memh = false,
-                           std::vector<uct_md_h> *recv_mds = NULL)
+                           bool send_prereg_memh = false,
+                           bool recv_prereg_memh = false)
     {
+        /* Both send and recv preregistered memory handles are not supported,
+         * since MDs shouldn't be resetted at least on one side to allow RKEY
+         * packing */
+        ASSERT_TRUE(!(send_prereg_memh & recv_prereg_memh));
+
         /* send multiple messages to test the protocol both before and after
          * connection establishment */
         for (int i = 0; i < num_iters; i++) {
             std::string sb(size, 'x');
             std::string rb(size, 'y');
             std::string shdr(hdr_size, 'x');
             std::string rhdr(hdr_size, 'y');
+            ucp_mem_h smemh(NULL);
             ucp_mem_h rmemh(NULL);
+            std::vector<uct_md_h> mds;
 
-            if (recv_prereg_memh) {
+            if (send_prereg_memh) {
```

Comment:
assert !(send_prereg_memh && recv_prereg_memh)

Reply:
we already have this check above